include/nbl/builtin/hlsl/emulated/float64_t.hlsl

```diff
@@ -0,0 +1,657 @@
+{\
+    return extractExponent(x.data);\
+}\
+\
+template<>\
+NBL_CONSTEXPR_INLINE_FUNC __VA_ARGS__ replaceBiasedExponent(__VA_ARGS__ x, typename unsigned_integer_of_size<sizeof(__VA_ARGS__)>::type biasedExp)\
+{\
+    return __VA_ARGS__(replaceBiasedExponent(x.data, biasedExp));\
+}\
+\
+template <>\
+NBL_CONSTEXPR_INLINE_FUNC __VA_ARGS__ fastMulExp2(__VA_ARGS__ x, int n)\
+{\
+    return __VA_ARGS__(replaceBiasedExponent(x.data, extractBiasedExponent(x) + uint32_t(n)));\
+}\
+\
+template <>\
+NBL_CONSTEXPR_INLINE_FUNC unsigned_integer_of_size<sizeof(__VA_ARGS__)>::type extractMantissa(__VA_ARGS__ x)\
+{\
+    return extractMantissa(x.data);\
+}\
+\
+template <>\
+NBL_CONSTEXPR_INLINE_FUNC uint64_t extractNormalizeMantissa(__VA_ARGS__ x)\
+{\
+    return extractNormalizeMantissa(x.data);\
+}\
+\
+
+#define DEFINE_BIT_CAST_SPEC(...)\
+template<>\
+NBL_CONSTEXPR_FUNC __VA_ARGS__ bit_cast<__VA_ARGS__, uint64_t>(NBL_CONST_REF_ARG(uint64_t) val)\
+{\
+__VA_ARGS__ output;\
+output.data = val;\
+\
+return output;\
+}\
+\
+template<>\
+NBL_CONSTEXPR_FUNC __VA_ARGS__ bit_cast<__VA_ARGS__, float64_t>(NBL_CONST_REF_ARG(float64_t) val)\
+{\
+__VA_ARGS__ output;\
+output.data = bit_cast<uint64_t>(val);\
+\
+return output;\
+}\
+\
+template<>\
+NBL_CONSTEXPR_FUNC uint64_t bit_cast<uint64_t, __VA_ARGS__ >(NBL_CONST_REF_ARG( __VA_ARGS__ ) val)\
+{\
+return val.data;\
+}\
+\
+template<>\
+NBL_CONSTEXPR_FUNC float64_t bit_cast<float64_t, __VA_ARGS__ >(NBL_CONST_REF_ARG( __VA_ARGS__ ) val)\
+{\
+return bit_cast<float64_t>(val.data);\
+}\
+\
+
+namespace impl
+{
+
+template<typename To, bool FastMath, bool FlushDenormToZero>
+struct static_cast_helper<To,emulated_float64_t<FastMath,FlushDenormToZero>,void>
+{
+    static_assert(is_scalar<To>::value);
+
+    using From = emulated_float64_t<FastMath,FlushDenormToZero>;
+     
+    static inline To cast(From v)
+    {
+        using ToAsFloat = typename float_of_size<sizeof(To)>::type;
+        using ToAsUint = typename unsigned_integer_of_size<sizeof(To)>::type;
+
+        if (emulated_float64_t_impl::isZero(v.data))
+            return 0;
+
+        if (is_same_v<To, float64_t>)
+            return To(bit_cast<float64_t>(v.data));
+
+        if (is_floating_point<To>::value)
+        {
+            const int exponent = ieee754::extractExponent(v.data);
+            if (!From::isFastMathSupported)
+            {
+                if (exponent > ieee754::traits<ToAsFloat>::exponentMax)
+                    return bit_cast<To>(ieee754::traits<ToAsFloat>::inf);
+                if (exponent < ieee754::traits<ToAsFloat>::exponentMin)
+                    return bit_cast<To>(-ieee754::traits<ToAsFloat>::inf);
+                if (tgmath::isNaN(v.data))
+                    return bit_cast<To>(ieee754::traits<ToAsFloat>::quietNaN);
+            }
+
+            const uint32_t toBitSize = sizeof(To) * 8;
+            const ToAsUint sign = ToAsUint(ieee754::extractSign(v.data) << (toBitSize - 1));
```

Comment:
`extractSign` should extract the sign and leave it where it is, unless you count on >>63 followed by <<63 to optimize to `&((1<<64)-1)` somehow